### PR TITLE
deletion of duplicate messages / originality enforcement mechanism

### PR DESCRIPTION
**Pull request by a-lexer**
Creativity must be enforced by an unrelenting, unyielding, uncompromising bot. Duplicate messages in #offtopic which do not match an excepted pattern (such as containing http or being <= 4 characters long) will be deleted.

I didn't know how to make the deletion message only visible to the single user. I also don't know if offtopic is a good channel to trial this. Finally, I don't know if this is good/fun/interesting, but it might at least provide a bit of amusement and healthy or unhealthy frustration for a few minutes.

Quite lightly tested :)

### CODE REVIEW

File: extensions/duplicate_message_policing.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    Deletes duplicate messages (excepting some). A duplicate message is simply a matching string
+    that has been seen before in the server (and not deleted).
+    """
+
+    nodelete_flag = "!"
+    # random rules. Probably worth thinking about this some more, if this bot function doesn't get deleted.
+    if (
+        (
+            event.channel_id != 813266892818219038 and os.environ.get("DEBUG") is None
```
Comment (flukiluke) [inline extensions/duplicate_message_policing.py]: Read channel id from environment

File: extensions/duplicate_message_policing.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        )  # channel id is for #offtopic
+        or event.content is None
+        or event.is_webhook
+        or event.content.startswith(nodelete_flag)
+        or event.is_bot
+        or not event.content
```
Comment (flukiluke) [inline extensions/duplicate_message_policing.py:35]: I think this also covers the `event.content is None` case?

File: extensions/duplicate_message_policing.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        or "http" in event.content  # allow links
+        or "@" in event.content  # allow mentions
+        or len(event.content) <= 4  # allow short messages
```
Comment (flukiluke) [inline extensions/duplicate_message_policing.py]: This still allows the string `lol` :(

File: extensions/duplicate_message_policing.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+This is designed to increase the effort required by users. For example, the string 'lol' can only be used
+once, and subsequent uses must get more creative (e.g. 'l0l', 'lolz', and so on).
+Eventually, some time in the future, all possible strings will have been said, and the database will need to be reset.
+
+Inspired by: https://blog.xkcd.com/2008/01/14/robot9000-and-xkcd-signal-attacking-noise-in-chat/
+"""
+import os, re
+import hikari, lightbulb
+import db
+import hashlib
```
Comment (flukiluke) [inline extensions/duplicate_message_policing.py]: `hashlib` is unused?

File: extensions/duplicate_message_policing.py (hunk 1 above)
Comment (a-lexer) [reply extensions/duplicate_message_policing.py]: nice, didn't know that could be done, I guess it's somewhere else in the code ;-)

File: extensions/duplicate_message_policing.py (hunk 3 above)
Comment (flukiluke) [reply extensions/duplicate_message_policing.py]: A little more consideration makes me think the appropriate check might be `<= 2`, to allow emoji and most common ascii emoticons, while still prohibiting the dreaded `lol`.

**Comment by acmahaja** [comment]
This is single handily the worst feature ever created in modern history

**Pull request merged**